### ISSUE TITLE
GitHub Copilot for commits

### ISSUE THREAD

**Issue by medallyon** [desktop/desktop#17439]
### The feature request

I think it would be neat to be able to generate commit messages (from a template) from my staged changes directly in Desktop. GitLens already does this using OpenAI. It could speed up workflows significantly.

### Proposed solution

Generate commit messages from a template that I'm able to set via Desktop.

**Comment by EwoutH** [comment]
I would love this as well!

It's now available in Visual Studio and VS Code, but I still really like the GitHub Desktop workflow. It would be great if it could be added.
- [Write Your Git Commits with GitHub Copilot](https://devblogs.microsoft.com/visualstudio/write-your-git-commits-with-github-copilot/)

**Comment by EwoutH** [comment]
JetBrains now also has generative AI commit messages in their 2023.3 releases: https://www.jetbrains.com/ai/

![generate-commit-message (1)](https://github.com/desktop/desktop/assets/15776622/ca5343c7-2d4f-42e7-942a-dab6a76dc2ac)

**Comment by 0xdevalias** [comment]
See also:

- https://github.com/desktop/desktop/issues/8698
- https://github.com/desktop/desktop/issues/17525

**Comment by jonlepage** [comment]
i would like also see this in github desktop !
It will be awesome feature, github desktop it my favorite ui for manage commit and branch.

**Comment by BrixsterPlanet** [comment]
Are there any plans to add support for this?

**Comment by emmadesilva** [comment]
I also think it's about time for GitHub Desktop to support AI commit message writing. Considering GitHub's focus on AI, it just makes sense.

**Comment by vatro** [comment]
This would be really cool + option to:
- make them [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) compliant
- make them < 50 chars long

**Comment by BennoCrafter** [comment]
Someone already working on it?

**Comment by Anselmoo** [comment]
> I also think it's about time for GitHub Desktop to support AI commit message writing. Considering GitHub's focus on AI, it just makes sense.

In this regard, a general chat function like [GitHub-Mobile-App](https://github.blog/news-insights/product-news/github-copilot-chat-in-github-mobile/) would be great for general research.

**Comment by gustavosett** [comment]
I want that so much... ;c

**Comment by BennoCrafter** [comment]
Hi all,

I'm currently working on this using Ollama, but I want to ensure the solution remains flexible and adaptable to other services in the future.
I'm open to feedback, suggestions, and collaboration with anyone interested in contributing. If you're interested in working on this or have ideas, feel free to reach out or share your thoughts here.

**Comment by BennoCrafter** [comment]
Edit:
I won't be able to work on this because the pull request draft was closed.

I’ve been working on a basic version of it, powered locally by Llama 3.2, and I wanted to give a quick showcase of it. It’s still missing some features, like a loading animation or a proper icon, but it’s functional!

Follow new changes here [#19866](https://github.com/desktop/desktop/pull/19866)

## What’s Missing:

- [ ] A loading animation
- [X] A proper icon
- [ ] More intelligent commit message generation (Better prompt)
- [ ] Different models (like copilot)

I’m still learning about prompt engineering and how to make the tool smarter. If anyone is experienced in this area and has some time to spare, feel free to post a better prompt for improving the tool’s suggestions. Any help would be greatly appreciated!

## Showcase
https://github.com/user-attachments/assets/4b37a0e8-1ac4-4387-9ba3-3e4ee1a9b22b

**Comment by ReessKennedy** [comment]
@BennoCrafter - You still using your version? You have a built DMG of your version I could try out/use? I am feeling like this would be so helpful I'd be willing to use yours until the Github core team implements this, if they do.

**Comment by BennoCrafter** [comment]
@ReessKennedy Try building it yourself based on my fork!

**Comment by EwoutH** [comment]
Really cool! Thanks for implementing this.

Would it be possible to add a field for custom instructions somewhere?

**Comment by fabiopedrosa** [comment]
~Great stuff, but I'm not seeing it show up. Is there instructions to somehow enable it?~
Had to enable it in the company policies:

![Image](https://github.com/user-attachments/assets/1246838e-0e0b-4e6b-9b6e-d28e93f1409b)

**Comment by ghost** [comment]
There needs to be a way to detect if most commit messages were generated with the help of AI. Because it's kind of reasonable to expect that users will just set it and forget it. No one wants to write each and every commit message so it follows that they will just automate generating commits instead of reviewing them eventually or maybe as soon as possible.

---

<details><summary>How to contact me</summary>
<p>

Manish

Check out my kind of oldfangled, kinda newfangled, kinda cool, inevitably thought-provoking profile. And the limited mentorship opportunities for your no matter the non business major you have chosen be it - Bachelor's or Master's or Doctorates or Post-Doctorates.
In short, if you want to do check out my Readme.md at https://github.com/ket395 or https://github.com/ket395/ket395. It's quite good, I declare. 

</p>
</details>

**Comment by LeMyst** [comment]
What about non-organization users?

**Comment by DanielGeek** [comment]
Awesome!

**Comment by josepcrespo** [comment]
Super useful feature to have the hability to generate commit messages directly from GitHub Desktop, I've been generating commit messages with third party AI tools since long time ago and, probably is my second favorite use of the AI. Thanks for implementing this feature.

**Comment by emmadesilva** [comment]
Amazing!

**Comment by LeMyst** [comment]
As said in another issue, it's available for non-organization user with Github Desktop Beta

**Comment by alvvaro** [comment]
Before updating, is this enabled by default? Can it be disabled completely on the app regardless of any organization settings? Does it honor the "Show Copilot" setting of the website?

**Comment by 0xdevalias** [comment]
@alvvaro When I updated, it appears to be enabled by default, but using it is a manual action.

From a quick skim I didn't see any settings within GitHub Desktop itself to remove/disable it, but I didn't look very hard.

I'm not sure if it honors the 'show copilot' setting of the website, but you could probably skim through the PR that added it to see:

- https://github.com/desktop/desktop/pull/20348

The docs/help page may also help answer your questions:

- https://docs.github.com/en/copilot/responsible-use-of-github-copilot-features/responsible-use-of-github-copilot-in-github-desktop
  - > Owners of organizations or enterprises with a GitHub Copilot Business or GitHub Copilot Enterprise subscription can decide whether to grant access to Copilot in GitHub Desktop for users in their organization or enterprise.
    > 
    > If you have a GitHub Copilot Free or GitHub Copilot Pro subscription, you now have access to Copilot in GitHub Desktop.
  - > Commit messages are only created when users request them manually.

**Comment by alvvaro** [comment]
![Image](https://github.com/user-attachments/assets/365b4ec5-946d-43cf-92df-eabac9f91b37)

Upon opening, GitHub Desktop queries the server about the status of this feature, I must assume it takes the value of the "Show Copilot" setting.

**Comment by 0xdevalias** [comment]
> Upon opening, GitHub Desktop queries the server about the status of this feature

@alvvaro You can see where `account.isCopilotDesktopEnabled` is used in the original PR here:

- https://github.com/desktop/desktop/pull/20348/files#diff-a3fa9bd8403f6e3c9e2500b461c11fe43d88bae5f48c3b48e2fc7946491bbecb

https://github.com/desktop/desktop/blob/c8c7413b2898faf7c4aaac43da75a8a86493a7d9/app/src/lib/feature-flag.ts#L109-L118

And `enableCommitMessageGeneration` is checked before rendering the button (`renderCopilotButton`):

https://github.com/desktop/desktop/blob/c8c7413b2898faf7c4aaac43da75a8a86493a7d9/app/src/ui/changes/commit-message.tsx#L873-L879

And within `_generateCommitMessage`:

https://github.com/desktop/desktop/blob/c8c7413b2898faf7c4aaac43da75a8a86493a7d9/app/src/lib/stores/app-store.ts#L5461-L5469

So yeah, I think your conclusion is likely correct.

**Comment by SAKSRiPRASERTSANG** [comment]
@SAKSRiPRASERTSANG

**Comment by YohannParis** [comment]
Why the new release remove that feature?
